Updated Chat.java

Mel98378 opened the pull request:
Combined the code for program 7 with the given layout

Mel98378 committed:
Updated Chat.java

Combined the code for program 7 with the given layout

lavender-aa commented:
i don't think the disconnect button is working properly, either application just hangs

lavender-aa commented:
for now i can still make a review

lavender-aa commented on program-7/Chat.java:
```diff
@@ -1,203 +1,285 @@
+ 
+     // KeyListener for Enter in chatbox
+     @Override public void keyTyped(KeyEvent e) {}
+     @Override public void keyPressed(KeyEvent e) {
```
textfields already send an event to `actionPerformed()` on enter keypress, i don't think keylistener is necessary

lavender-aa commented on program-7/Chat.java:
```diff
@@ -1,203 +1,285 @@
+         }
+     }
+ 
+     void stop() {
```
i'm having a hard time figuring out why but the app gets stuck after disconnecting, i think it may have something to do with the way you wrote the thread? he never taught us anonymous functions anyways, pretty sure all thread related stuff is supposed to happen in `run()`

lavender-aa commented on program-7/Chat.java:
```diff
@@ -1,203 +1,285 @@
+     private boolean isServer = false;
+ 
+     public Chat() {
+         super("Chat");
```
i might be missing something, but i don't see why this is necessary

lavender-aa commented on program-7/Chat.java:
```diff
@@ -1,203 +1,285 @@
+     private void closeConnection() {
+         try {
+             if (reader != null) reader.close();
+             if (writer != null) writer.close();
```
i think part of the disconnect issue might be that you never send anything to the other instance to tell it that the connection is gone, so it gets null where it expects something maybe?

lavender-aa reviewed (changes requested):
has a few issues, hard to do a full review on everything.

some of this code is strange-looking, did you take help from chatgpt? it seems like it differs a bunch from what we've been learning in class

Mel98378 committed:
fixed problems

thread related stuff happens in run(), disconnect button is working properly

lavender-aa commented on program-7/Chat.java:
```diff
@@ -1,203 +1,295 @@
+     private void startServer() {
+         try {
+             int port = Integer.parseInt(portTF.getText());
+             serverSocket = new ServerSocket(port);
```
server socket has no timeout (thinking something like 1 minute would be fine)

lavender-aa commented on program-7/Chat.java:
```diff
@@ -1,203 +1,295 @@
+             String host = hostTF.getText();
+             int port = Integer.parseInt(portTF.getText());
+             statusTA.setText("Connecting to " + host + ":" + port + "...");
+             socket = new Socket(host, port);
```
client socket has no timeout (thinking 30s would be good)

lavender-aa commented on program-7/Chat.java:
```diff
@@ -1,203 +1,295 @@
+                 }
+ 
+                 // Disable the send button and update the status
+                 sendBTN.setEnabled(false);
```
i don't know if it's necessary but i think it would be a good idea to clear the chats on disconnect

lavender-aa reviewed (changes requested):
disconnect button does work, just a few smaller things that seem like they won't take too long

lavender-aa commented:
oh also you never answered, did you use chatgpt to write some of this?

Mel98378 committed:
Server/Client Socket timeout and clear when disconnected

Added timeouts for Server and Client sockets. Clears chat when disconnected. And yeah i used Chatgpt to help with some of the problems i was having

lavender-aa commented on program-7/Chat.java:
```diff
@@ -1,203 +1,296 @@
+                     listenThread.join(100);
+                 }
+ 
+                 chatTA.setText("Disconnected. Chat cleared.\n");
```
this should most likely go inside the status textarea, unless you also clear the chats on reopen (in `startServer()` and `connectToServer()`

lavender-aa commented on program-7/Chat.java:
```diff
@@ -1,203 +1,296 @@
+     private void startServer() {
+         try {
+             int port = Integer.parseInt(portTF.getText());
+             serverSocket = new ServerSocket(port);
```
after server times out once, this will always throw an `IOException`. 

calling `closeConnection()` inside of the `SocketTimeoutException` catch block seems to fix the issue

lavender-aa reviewed (changes requested):
more small changes

lavender-aa commented:
for the two places you create new threads with anonymous functions, you can just remove the thread wrapper. the code inside doesn't need to be run inside the thread

also, if you could, please remove the leading space on every line

Mel98378 committed:
small changes/removed anonymous functions

Chat cleared message is displayed in status area now, fixed IOException problem, removed anonymous functions, and removed leading spaces.

lavender-aa reviewed (approved):
works well, i may add some comments describing why extra threads need to be created
